### PR TITLE
Improve performance for loading files

### PR DESCRIPTION
**Pull request by tobiasdiez**
- Performance improvements around groups
- Remove detection of duplicate ID when inserting entries (it's not really possible to create two entries with the same except if you use `setId` manually)
- Remove detection of duplicate bibtex keys when opening a file (the result was not used and we have a integrity check for it)
- Use EnumMap instead of HashMap to cache fields as keywords (which is only used for Keyword and Groups fields anyway)
- Fixes https://github.com/JabRef/jabref/issues/6329 where latex code was displayed in the maintable
- Lazy init of source tab

In summary, we are now at very reasonable numbers:
- Reading files is about 1k entries per second (with 30% spent on initialing the EventBus)
- Groups membership counter is still expensive to calculate but kind of ok (30 sec for 1k groups x 12k entries = 12 mil checks)
- 7s load of an empty JabRef instance
- 1gb ram with two databases of together 15k entries
  
(all numbers are measured under profiling, thus they are worse than under actual production).

I think that concludes my series of performance improvements for now. 

- [ ] Change in CHANGELOG.md described (if applicable)
- [ ] Tests created for changes (if applicable)
- [ ] Manually tested changed features in running JabRef (always required)
- [ ] Screenshots added in PR description (for UI changes)
- [ ] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, submitted a pull request to the documentation repository.

### CODE REVIEW

**Comment by Siedlerchr** [comment]
Regarding, entries with the same id: What happens when you have a shared database? @koppor

File: src/main/java/org/jabref/model/util/ListUtil.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
+package org.jabref.model.util;
+
+import java.util.function.Predicate;
+
+public class ListUtil {
+    public static <T> boolean anyMatch(Iterable<T> list, Predicate<T> predicate) {
```
[the remark this answers is not in the document]
Comment (tobiasdiez) [reply src/main/java/org/jabref/model/util/ListUtil.java:6]: It's about a factor two faster than the streams solution.

**Comment by tobiasdiez** [comment]
https://github.com/JabRef/jabref/blob/3a899d84311abd98c6cd0579ff2d7b9bc90ebec1/src/main/java/org/jabref/gui/collab/EntryChangeViewModel.java#L48-L50
That's the only instance where `setId` is invoked and since the old entry is removed shortly before, there is really no chance of duplicate id.

File: src/main/java/org/jabref/gui/entryeditor/SourceTab.java [hunk 2]
```diff
@@ -172,6 +169,7 @@ public Point2D getTextLocation(int offset) {
     }
 
     private void setupSourceEditor() {
+        codeArea = new CodeArea();
```
Comment (calixtus) [inline src/main/java/org/jabref/gui/entryeditor/SourceTab.java:172]: What about `if (codeArea == null) { codeArea == new CodeArea() }`?

**Comment by koppor** [comment]
> Regarding, entries with the same id: What happens when you have a shared database? @koppor

In general, we have the `sharedId` at the communication with the server - to ensure that two users always have the same id for the same entry.

File: .idea/runConfigurations/JabRef_Main.xml [hunk 3]
```diff
@@ -2,6 +2,7 @@
   <configuration default="false" name="JabRef Main" type="Application" factoryName="Application" singleton="false">
     <option name="MAIN_CLASS_NAME" value="org.jabref.JabRefMain" />
     <module name="JabRef.main" />
+    <option name="REDIRECT_INPUT" value="true" />
```
Comment (koppor) [inline .idea/runConfigurations/JabRef_Main.xml:5]: This is not related to this PR, is it?

File: src/main/java/org/jabref/gui/maintable/BibEntryTableViewModel.java [hunk 4]
```diff
@@ -41,10 +39,10 @@
     private final ObjectBinding<Map<Field, String>> linkedIdentifiers;
     private final ObservableValue<List<AbstractGroup>> matchedGroups;
 
-    public BibEntryTableViewModel(BibEntry entry, BibDatabaseContext database) {
+    public BibEntryTableViewModel(BibEntry entry, BibDatabaseContext database, MainTableNameFormatter nameFormatter) {
```
Comment (koppor) [inline src/main/java/org/jabref/gui/maintable/BibEntryTableViewModel.java:42]: 👍

File: src/main/java/org/jabref/logic/importer/ParserResult.java [hunk 5]
```diff
@@ -35,7 +34,7 @@ public ParserResult() {
     }
 
     public ParserResult(Collection<BibEntry> entries) {
-        this(BibDatabases.createDatabase(BibDatabases.purgeEmptyEntries(entries)));
+        this(new BibDatabase(BibDatabases.purgeEmptyEntries(entries)));
```
Comment (koppor) [inline src/main/java/org/jabref/logic/importer/ParserResult.java:37]: This is a kind of taste, isn't it? - Would assume that the Builder pattern would be preferred.

File: src/main/java/org/jabref/model/database/BibDatabase.java [hunk 6]
```diff
@@ -169,46 +169,35 @@ public boolean containsEntryWithId(String id) {
...
+     * Inserts the entry.
      *
-     * @param entry BibEntry to insert into the database
-     * @return false if the insert was done without a duplicate warning
-     * @throws KeyCollisionException thrown if the entry id ({@link BibEntry#getId()}) is already  present in the database
+     * @param entry entry to insert
      */
-    public synchronized boolean insertEntry(BibEntry entry) throws KeyCollisionException {
-        return insertEntry(entry, EntriesEventSource.LOCAL);
+    public synchronized void insertEntry(BibEntry entry) {
```
Comment (koppor) [inline src/main/java/org/jabref/model/database/BibDatabase.java:176]: Finally 🎉

File: src/main/java/org/jabref/model/entry/field/InternalField.java [hunk 7]
```diff
@@ -12,7 +12,6 @@
 public enum InternalField implements Field {
     OWNER("owner"),
     TIMESTAMP("timestamp", FieldProperty.DATE),
-    GROUPS("groups"),
```
Comment (koppor) [inline src/main/java/org/jabref/model/entry/field/InternalField.java:15]: Wasn't the samentics of "InternalField" that is a field introduced by JabRef? (The class should be ranamed to `JabRefSuperPowerFields` (when going extreme).

`OWNER` and `TIMESTAMP` are "normal" BibTeX fields - just introdced by JabRef.

If we move `GROUPS` to `StandardField`, we should move the other two, too.

@stefan-kolb What's your take on this?

File: src/main/java/org/jabref/model/entry/field/StandardField.java [hunk 8]
```diff
@@ -124,7 +124,10 @@
     YEARFILED("yearfiled"),
     MR_NUMBER("mrnumber"),
     XDATA("xdata", FieldProperty.MULTIPLE_ENTRY_LINK),
-    XREF("xref", FieldProperty.SINGLE_ENTRY_LINK);
+    XREF("xref", FieldProperty.SINGLE_ENTRY_LINK),
+
+    // JabRef-specific fields
```
Comment (koppor) [inline src/main/java/org/jabref/model/entry/field/StandardField.java:129]: The comment also states, it should left in `InternalFields`.

File: src/main/java/org/jabref/model/groups/WordKeywordGroup.java [hunk 9]
```diff
@@ -148,4 +136,59 @@ public int hashCode() {
                 keywordSeparator,
                 onlySplitWordsAtSeparator);
     }
+
+    interface SearchStrategy {
+        boolean contains(BibEntry entry);
+    }
+
+    class StringSearchStrategy implements SearchStrategy {
```
Comment (koppor) [inline src/main/java/org/jabref/model/groups/WordKeywordGroup.java:144]: I would put the class in a seperate `.java` class. This would also enable testing with `...Test.java` files.

File: src/main/java/org/jabref/model/util/ListUtil.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (koppor) [reply src/main/java/org/jabref/model/util/ListUtil.java:6]: Can you just add a JavaDoc, please?

Would it be possible to add a performance test? So that if we replace this by something with promising performance (e.g., Goldman Sach's Collections), we can easily check which one is faster.

File: src/main/java/org/jabref/model/entry/field/InternalField.java (hunk 7 above)
Comment (stefan-kolb) [reply src/main/java/org/jabref/model/entry/field/InternalField.java:15]: Probably `JabrefCustomFields`, `CustomFields`?

File: .idea/runConfigurations/JabRef_Main.xml (hunk 3 above)
Comment (tobiasdiez) [reply .idea/runConfigurations/JabRef_Main.xml:5]: Not directly...was added by intellj automatically.

File: src/main/java/org/jabref/logic/importer/ParserResult.java (hunk 5 above)
Comment (tobiasdiez) [reply src/main/java/org/jabref/logic/importer/ParserResult.java:37]: Builder pattern for a constructor with one argument seems a bit overkill, don't you think?

File: src/main/java/org/jabref/model/entry/field/InternalField.java (hunk 7 above)
Comment (tobiasdiez) [reply src/main/java/org/jabref/model/entry/field/InternalField.java:15]: I now moved all three (group, owner, timestamp) to standard fields. They need to be grouped with the other fields in one enum to use the power of the enummap. Now the "StandardField" refers to a field that is a normal field for JabRef.

File: src/main/java/org/jabref/model/groups/WordKeywordGroup.java (hunk 9 above)
Comment (tobiasdiez) [reply src/main/java/org/jabref/model/groups/WordKeywordGroup.java:144]: This is more an implementation detail which shouldn't exposed to the outside world in my opinion. Moreover, private variables are used which makes extraction awkward.

**Comment by tobiasdiez** [comment]
Thanks for the reviews. I (hopefully) addressed each concern and updated the code accordingly.

**Comment by calixtus** [comment]
Great work!

I recently thought about another way to make user experience more responsive: The database tab could be opened, before the parser starts working. This would need some more refactoring (maybe moving the call to the import routines inside the tab constructor?). This would be a project in the mid future.

**Comment by tobiasdiez** [comment]
That's indeed a very good suggestion @calixtus. Let's keep it in mind.

**Comment by koppor** [comment]
@calixtus Could you please open an issue for it. Seems to be a good-first-issue? (And link this PR from the description there)
